### PR TITLE
Server join queue

### PR DESCRIPTION
**Pull request by Morb0**
## About the PR 
Implementation of a simple queue for fair connection of players to the desired server.
The queue keeps the player's connection state in `Connected` and only if `total number of players - number in queue` becomes lower than `soft_max_players` the system changes the player's state to `InGame`.

**The queue is disabled in CVar by default!**

When the queue is activated setting `soft_max_players` shows only how many players can join the game, all the rest stand in queue, but admins and those who were in a round still bypass this restriction.
Also, if the queue was disabled though cvar during the game, everyone who was in the queue will be kicked from the server.
Server API for hub gives only players in game, not counting that stay in queue.
On successful connect, *ding sound played.

Engine PR https://github.com/space-wizards/RobustToolbox/pull/3531 not required, but will fix error when disconnecting a user in the queue:
`[ERRO] PVS: Attempted to remove player from pvsCollection, but they were already removed? Session:localhost@Morty`

**Screenshots**

image
Demonstration: https://streamable.com/i7yav7

**Changelog**

### CODE REVIEW

**Comment by ElectroJr** [comment]
I know the consensus for wizard's den servers is to not have a queue, but I guess having the option available for downstreams who want the feature is fine.

However, even with that being the case I think any sort of queuing system like this would benefit significantly from having some system that automatically starts a vote for the queued players saying something like "Hey, 10 of you are queued up, how about you join [some underpopulated server] instead?".

**Comment by Morb0** [comment]
> However, even with that being the case I think any sort of queuing system like this would benefit significantly from having some system that automatically starts a vote for the queued players saying something like "Hey, 10 of you are queued up, how about you join [some underpopulated server] instead?". Though I guess that can also always be done in a separate PR.

I would rather display a short list of unfilled servers below the queue information, since servers have their own criteria, such as language, rp level, etc.
But in essence it will be copying code from the launcher which will make requests to literally every server to get its online...
And I think people know that they can go to another server and not stand in queue... a lot of people do.

File: Content.Server/JoinQueue/JoinQueueManager.cs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        var queueContains = _queue.Count > 0;
+        if ((!_isEnabled || haveFreeSlot) && queueContains)
+        {
+            var session = _queue.First();
+            _queue.Remove(session);
+            SendToGame(session);
+        }
+
+        SendUpdateMessages();
+    }
```
Comment (metalgearsloth) [inline Content.Server/JoinQueue/JoinQueueManager.cs:99]: Is this better off just using an actual Queue<T>?

**Comment by moonheart08** [comment]
> I know the consensus for wizard's den servers is to not have a queue, but I guess having the option available for downstreams who want the feature is fine.
> 
> However, even with that being the case I think any sort of queuing system like this would benefit significantly from having some system that automatically starts a vote for the queued players saying something like "Hey, 10 of you are queued up, how about you join [some underpopulated server] instead?". Though I guess that can also always be done in a separate PR.

I'm pinning this as a requirement for integrating queues upstream. People very often will sit around and wait instead of playing and forcing them into another server this way will improve traffic flow. I don't trust players to self-sort with a queue in place and don't want to harm the global server population by having people bunch up.

File: Content.Server/JoinQueue/JoinQueueManager.cs (hunk 1 above)
Comment (Morb0) [reply Content.Server/JoinQueue/JoinQueueManager.cs:99]: Actual Queue can't `.Remove` disconnected players

**Comment by Morb0** [comment]
Only caused more problems
return to autoclickers 🌵

**Pull request closed without merging**